src/client/Microsoft.Identity.Client/Instance/Region/RegionManager.cs

```diff
@@ -0,0 +1,218 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.Identity.Client.TelemetryCore.Internal.Events;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.Region
+{
+    internal sealed class RegionManager : IRegionManager
```

Comment:
This needs a careful review.

Reply:
it looks good so far, i would need to run a unit test or manually test, which i will do later.

---
In reply to: [602424586](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2509#discussion_r602424586) [](ancestors = 602424586)

Reply:
lgtm. had to read the code a few times to understand the flow

---
In reply to: [602521119](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2509#discussion_r602521119) [](ancestors = 602521119,602424586)